### PR TITLE
[Quant][fx] Enable FX static quantization for LSTM

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85068

**Summary:** This commit enables the custom module LSTM path for
FX graph mode static quantization. This has the same flow as eager
mode, which was already previously supported:

```
     torch.nn.LSTM
           | (prepare_fx)
           v
torch.ao.nn.quantizable.LSTM
           | (convert_fx)
           v
 torch.ao.nn.quantized.LSTM
```

The main reason why custom module LSTM is not supported in FX
graph mode quantization today is because its inputs and outputs
are nested tuples, and existing constructs such as observers,
"quantize" nodes, and "dequantize" nodes do not understand how
to handle complex structures.

Note that the approach taken in this commit is only intended to
be a short-term solution highly tailored to the input and output
formats of custom module LSTM. In the future, for the longer-term
solution, we should design a more general QConfig that allows users
to specify complex input and output formats, and enable FX graph
mode quantization to understand arbitrary nested structures and
automatically infer how to transform the graph accordingly.

**Context:**

Today, in FX graph mode static quantization, custom modules are
assumed to have quantized inputs and quantized outputs, with the
exact dtypes derived from the associated QConfig (default quint8).
Since custom modules are currently not handled through the reference
model flow, their observer replacement logic are a little different
from normal operators:

```
# (1) Original model
input -> custom_module -> output

# (2) Observed model (after prepare)
input -> obs0 -> custom_module -> obs1 -> output

# (3) Quantized model (after convert)
input -> quant -> quantized_custom_module -> dequant -> output
```

In the last step, input observers are replaced with "quantize"
and output observers are replaced with "dequantize", in contrast
to other non-custom-module patterns where observers are replaced
with "quantize-dequantize" pairs instead. Note that, conceptually,
the output observer `obs1` is really just a DeQuantStub, since no
observation is actually needed.

**Custom module LSTM:**

The reason why custom module LSTM cannot be handled in the same
way is because, unlike other custom modules, its inputs and outputs
are nested tuples instead of single tensors. This is how the existing
custom module code would try to handle LSTMs:

```
# (1) Original model
# input format: (input, (hidden0, hidden1))
# output format:  (output, (hidden0, hidden1))
 input -> lstm -> output
hidden0 -/    \-> hidden0
hidden1 -/    \-> hidden1

# (2) Observed model (after prepare)
 input -> obs0 -> lstm -> obs1  # fails
        hidden0 -/  # missing observer
        hidden1 -/  # missing observer
```

However, this fails today because 1) we assume there is only one input
to the custom module, and so we never end up quantizing `hidden0` and
`hidden1`, and 2) the output observer `obs1` is fed a tuple, which it
does not understand how to handle.

**Short-term fix:**

This commit addresses the above by specifically handling the input
and output structures used by custom module LSTM. For the inputs,
we manually insert observers for `hidden0` and `hidden1` to ensure
all input tensors are quantized.

For the outputs, we split the tuple into its internal nodes, attach
a DeQuantStub to each node, and recombine these DeQuantStubs
according to the original structure. Finally, we must also reroute
consumers of the original LSTM tuple (and its internal nodes, e.g.
`lstm[0]`) to these DeQuantStubs:

```
# (1) Original model
 input -> lstm -> output -> linear0
hidden0 -/    \-> hidden0 -> linear1
hidden1 -/    \-> hidden1 -> linear2

# (2) Observed model (after prepare)
 input -> obs0 -> lstm -> output -> dqstub -> linear0 -> obs3
hidden0 -> obs1 -/    \-> hidden0 -> dqstub -> linear1 -> obs4
hidden1 -> obs2 -/    \-> hidden1 -> dqstub -> linear2 -> obs5

# (3) Reference model (after convert)
 input -> quant -> qlstm -> output -> dequant -> linear0 -> quant -> dequant
hidden0 -> quant -/    \-> hidden0 -> dequant -> linear1 -> quant -> dequant
hidden1 -> quant -/    \-> hidden1 -> dequant -> linear2 -> quant -> dequant

# (4) Quantized model (after lowering)
 input -> quant -> qlstm -> output -> quantized_linear0 -> dequant
hidden0 -> quant -/    \-> hidden0 -> quantized_linear1 -> dequant
hidden1 -> quant -/    \-> hidden1 -> quantized_linear2 -> dequant
```

Note that we choose to insert DeQuantStubs here instead of observers
because these will ultimately be replaced by "dequantize" nodes. This
matches the general custom module behavior, where output observers
are replaced only with "dequantize" nodes (as opposed to the normal
"quantize-dequantize" pair), since custom module outputs are assumed
to already be quantized. Using DeQuantStubs instead of observers also
simplifies the "dequantize" insertion logic. In the future, we should use
DeQuantStubs in place of output observers for custom modules in general.

**Implementation:**

In prepare, we make two special cases:

(1) When inserting output observers, we instead break the LSTM output
tuple into its internal nodes, insert a DeQuantStub after each node,
and recombine the tuple according to the original format
(in `_insert_dequant_stubs_for_custom_module_lstm_output`). This
diverges from, but is conceptually the same as, how we insert "special"
output observers that will be converted to "dequantize" nodes (during
convert) today for other custom modules. In the future, we should just
insert DeQuantStubs instead of these "special" observers for custom
modules in general.

(2) When inserting input observers for a node, we check whether it is
a consumer of LSTM by traversing up the subgraph created in (1)
(see `_maybe_get_custom_module_lstm_from_node_arg`).
Once we identified what the previous node is, we reuse the existing code
in `maybe_insert_input_observer_for_arg_or_kwarg` to decide whether or
not to insert the input observer based on the output dtype of the
previous node.

As for convert:

(1) When converting the custom module, we change
`quantize - dequantize - custom_module` to `quantize - custom_module`
for all inputs of custom module LSTM, including the hidden inputs (in
`convert_custom_module`). This is consistent with how we do the same
for custom module inputs today, except it handles more than one input.

(2) We replace DeQuantStubs with "dequantize" nodes. This is not
really the special case as we plan to do this for all custom modules
in the future.

**Test plan:**
python test/test_quantization.py TestQuantizeFx.test_static_lstm
python test/test_quantization.py
TestQuantizeFx.test_static_lstm_consume_tuple

**Reviewers:** jerryzh168, vkuzo

**Subscribers:** jerryzh168, vkuzo

### CODE REVIEW

File: test/quantization/fx/test_quantize_fx.py [hunk 1]
```diff
@@ -4082,6 +4082,57 @@ def forward(self, inputs: torch.Tensor, state: List[torch.Tensor]):
...
+            def __init__(self):
+                super().__init__()
+                self.lstm = nn.LSTM(50, 50, 1)
+
+            def forward(self, inputs: torch.Tensor, h0: torch.Tensor, c0: torch.Tensor):
+                return self.lstm(inputs, (h0, c0))
+
+        m = MyModel()
+        qconfig_mapping = get_default_qconfig_mapping()
+        prepare_custom_config = PrepareCustomConfig() \
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:4099]: nit: maybe we can have a get_default_prepare_custom_config as well (can be in a separate PR)

File: test/quantization/fx/test_quantize_fx.py [hunk 2]
```diff
@@ -4082,6 +4082,57 @@ def forward(self, inputs: torch.Tensor, state: List[torch.Tensor]):
...
+            .set_float_to_observed_mapping(torch.nn.LSTM, torch.ao.nn.quantizable.LSTM)
+        convert_custom_config = ConvertCustomConfig() \
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:4101]: same for this one

File: test/quantization/fx/test_quantize_fx.py [hunk 3]
```diff
@@ -4082,6 +4082,57 @@ def forward(self, inputs: torch.Tensor, state: List[torch.Tensor]):
             if n.target == "lstm":
                 self.assertEqual(type(n.args[1]), tuple)
 
+    def test_static_lstm(self):
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:4085]: I think we can add another test that uses lstm in the middle of the model and check if it works with surrounding operators or not

File: torch/ao/quantization/fx/utils.py [hunk 4]
```diff
@@ -640,3 +640,12 @@ def get_skipped_module_name_and_classes(
         skipped_module_classes += get_custom_module_class_keys(prepare_custom_config.float_to_observed_mapping)
 
     return skipped_module_names, skipped_module_classes
+
+def _is_custom_module_lstm(node: Node, modules: Dict[str, torch.nn.Module]) -> bool:
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/utils.py:644]: nit: modules --> named_modules

File: torch/ao/quantization/fx/prepare.py [hunk 5]
```diff
@@ -1279,6 +1285,7 @@ def insert_observers_for_model(
 
                             is_observer_in_same_graph_ = is_observer_in_same_graph(node, modules, node_name_to_target_dtype)
 
+                        if maybe_output_obs_node is not None or _is_custom_module_lstm(node, modules):
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/prepare.py:1288]: what does this check mean?

Comment (andrewor14) [reply torch/ao/quantization/fx/prepare.py:1288]: Since we no longer return an output observer for LSTM in L1261, we can skip updating the graph to point to the new observer (L1264-L1284), but we still need to swap the custom module, so we do a check for LSTM here.

Comment (andrewor14) [reply torch/ao/quantization/fx/prepare.py:1288]: But actually maybe this line should be right above L1297 instead. Let me move it

File: torch/ao/quantization/fx/convert.py [hunk 6]
```diff
@@ -456,6 +458,23 @@ def convert_weighted_module(
         parent_name, name = _parent_name(node.target)
         setattr(modules[parent_name], name, ref_qmodule)
 
+def _remove_previous_dequantize_in_custom_module(node: Node, prev_node: Node, graph: Graph):
+    """
+    Change the connection for custom module, we'll change the input
+    of custom module node to quantize node:
+    Before: quantize - dequantize - custom - module
```
Comment (vkuzo) [inline torch/ao/quantization/fx/convert.py:465]: nit: could we clarify what this graph is doing? Is `custom - module` a single node or two nodes?

File: torch/ao/quantization/fx/convert.py [hunk 7]
```diff
@@ -686,7 +708,8 @@ def replace_observer_with_quantize_dequantize_node(
 
     # this is a temporary hack for custom module, we may want to implement
     # this properly after the custom module class design is finalized
-    def replace_observer_with_dequantize_node(node: Node, graph: Graph):
+    # TODO: Use this only for replacing DeQuantStubs; currently it's also used for replacing observers
+    def replace_with_dequantize_node(node: Node, graph: Graph):
```
Comment (vkuzo) [inline torch/ao/quantization/fx/convert.py:712]: nit: `replace_observer_or_dequant_stub_with_dequantize_node` would be a more descriptive name

File: torch/ao/quantization/fx/convert.py (hunk 6 above)
Comment (andrewor14) [reply torch/ao/quantization/fx/convert.py:465]: Sure. I just moved this comment but I can clarify it. I think `custom - module` is supposed to be a single node

File: torch/ao/quantization/fx/prepare.py [hunk 8]
```diff
@@ -433,7 +435,13 @@ def get_arg_target_dtype_as_output(
     argument in quantized graph will match what is specified by the qconfig
     """
     assert isinstance(arg, Node)
-    if is_activation_post_process_node(arg, modules):
+    # Custom module LSTM output is a tuple that we broke down into the internal nodes in order
```
Comment (vkuzo) [inline torch/ao/quantization/fx/prepare.py:438]: this makes sense if I read the PR summary, but not by itself.  Can we add some more context here?

File: torch/ao/quantization/fx/utils.py [hunk 9]
```diff
@@ -640,3 +641,220 @@ def get_skipped_module_name_and_classes(
...
+
+def _get_custom_module_lstm_from_node_arg(
+    arg: Node,
+    named_modules: Dict[str, torch.nn.Module],
+) -> Optional[Node]:
+    """
+    Given an argument of a node, if the argument refers to the path through which the node
+    is a consumer of custom module LSTM, return the custom module LSTM node, or None otherwise.
+
+    This is used to determine whether to insert input observers for a node. Conceptually,
```
Comment (vkuzo) [inline torch/ao/quantization/fx/utils.py:809]: is this function used for LSTM inputs, LSTM outputs, or both?

File: torch/ao/quantization/fx/utils.py [hunk 10]
```diff
@@ -640,3 +641,220 @@ def get_skipped_module_name_and_classes(
...
+    nodes_added = [output, output_dq, hidden, hidden_dq, hidden0, hidden0_dq, hidden1, hidden1_dq]
+    nodes_to_skip = set(node_to_replacement.keys()).union(nodes_added)
+    for original_node, replacement in node_to_replacement.items():
+        for user in list(original_node.users.keys()):
+            if user not in nodes_to_skip:
+                user.replace_input_with(original_node, replacement)
+
+    return lstm_output_dq
+
+def _get_custom_module_lstm_from_node_arg(
```
Comment (vkuzo) [inline torch/ao/quantization/fx/utils.py:801]: nit: `_maybe_get_custom_module_lstm_from_node_arg`?

File: torch/ao/quantization/fx/utils.py [hunk 11]
```diff
@@ -640,3 +641,220 @@ def get_skipped_module_name_and_classes(
...
     return skipped_module_names, skipped_module_classes
+
+def _is_custom_module_lstm(node: Node, named_modules: Dict[str, torch.nn.Module]) -> bool:
+    """
+    Return whether this refers to the custom module LSTM flow.
+    """
+    if node.op != "call_module" or str(node.target) not in named_modules:
+        return False
+    mod = named_modules[str(node.target)]
+    return isinstance(mod, torch.nn.LSTM) or isinstance(mod, torch.ao.nn.quantizable.LSTM)
```
Comment (vkuzo) [inline torch/ao/quantization/fx/utils.py:652]: `isinstance(mod, torch.nn.LSTM)` sounds too permissive, is there a way to restrict it to modules going through the custom module flow only?

File: torch/ao/quantization/fx/utils.py [hunk 12]
```diff
@@ -640,3 +641,220 @@ def get_skipped_module_name_and_classes(
...
+
+def _is_getitem_node(node: Node) -> bool:
+    """
+    Return whether `node` refers to a `getitem` call_function node.
+    """
+    return node.op == "call_function" and node.target == operator.getitem
```
Comment (vkuzo) [inline torch/ao/quantization/fx/utils.py:658]: why not keep it inlined?

File: torch/ao/quantization/fx/utils.py [hunk 13]
```diff
@@ -640,3 +641,220 @@ def get_skipped_module_name_and_classes(
...
+    """
+    if node.op != "call_module" or str(node.target) not in named_modules:
+        return False
+    return isinstance(named_modules[str(node.target)], DeQuantStub)
+
+def _is_tuple_node(node: Node) -> bool:
+    """
+    Return whether `node` refers to a `tuple` call_function node.
+    """
+    return node.op == "call_function" and node.target == tuple
```
Comment (vkuzo) [inline torch/ao/quantization/fx/utils.py:672]: why not keep it inlined?

File: torch/ao/quantization/fx/utils.py [hunk 14]
```diff
@@ -640,3 +641,220 @@ def get_skipped_module_name_and_classes(
...
+    """
+    return node.op == "call_function" and node.target == operator.getitem
+
+def _is_dequant_stub(node: Node, named_modules: Dict[str, torch.nn.Module]) -> bool:
+    """
+    Return whether `node` refers to a `DeQuantStub` call_module node.
+    """
+    if node.op != "call_module" or str(node.target) not in named_modules:
+        return False
+    return isinstance(named_modules[str(node.target)], DeQuantStub)
```
Comment (vkuzo) [inline torch/ao/quantization/fx/utils.py:666]: nit: I think the isinstance check should be inlined.  Otherwise people have to create a new util function for every type they want to check against.

Something like

```
from torch.ao.quantization.utils import getattr_from_fqn

mod = getattr_from_fqn(model, node.target)
if isinstance(mod, DeQuantStub):
    ...
```

would be cleaner

File: test/quantization/fx/test_quantize_fx.py [hunk 15]
```diff
@@ -4082,6 +4082,126 @@ def forward(self, inputs: torch.Tensor, state: List[torch.Tensor]):
...
+        self.checkGraphModuleNodes(model, expected_node_occurrence=prepare_node_occurrence)
+        model(*example_inputs)
+
+        model = convert_fx(model, convert_custom_config=convert_custom_config)
+        self.checkGraphModuleNodes(model, expected_node_occurrence=convert_node_occurrence)
+        model(*example_inputs)
+
+    def test_static_lstm(self):
+        """
+        Test static quantization for LSTM.
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:4106]: this doesn't provide any extra information compared to the name, maybe a more descriptive comment would be more helpful

File: test/quantization/fx/test_quantize_fx.py [hunk 16]
```diff
@@ -4082,6 +4082,126 @@ def forward(self, inputs: torch.Tensor, state: List[torch.Tensor]):
...
+                x = self.lstm(inputs, (h0, c0))
+                x = self.module_after_lstm(x)  # consume tuple (output, (hidden0, hidden1))
+                return x
+
+        class ConsumeHiddenTuple(ConsumeWholeTuple):
+            def forward(self, inputs: torch.Tensor, h0: torch.Tensor, c0: torch.Tensor):
+                x = self.lstm(inputs, (h0, c0))
+                x = self.module_after_lstm(x[1])  # consume tuple (hidden0, hidden1)
+                return x
+
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:4170]: maybe add a test to consume x[0] as well? just for completeness

File: torch/ao/quantization/fx/utils.py [hunk 17]
```diff
@@ -640,3 +641,220 @@ def get_skipped_module_name_and_classes(
...
+      lstm -> consumer
+
+    In practice, however, custom module LSTM outputs a tuple (output, (hidden0, hidden1)) with
+    DeQuantStubs attached to each internal node (see `_insert_dequant_stubs_for_custom_module_lstm_output`).
+    This tuple can be consumed in one of four ways:
+
+      lstm -> getitem -> DeQuantStub -> consumer                       # consume lstm[0]
+      lstm -> getitem -> getitem -> DeQuantStub -> tuple -> consumer   # consume lstm[1]
+      lstm -> getitem -> getitem -> DeQuantStub -> consumer            # consume lstm[1][0] or lstm[1][1]
+      lstm -> getitem -> DeQuantStub -> tuple -> consumer              # consume lstm
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/utils.py:822]: this is consuming the whole lstm output? why do we have getitem here?

File: test/quantization/fx/test_quantize_fx.py (hunk 15 above)
Comment (andrewor14) [reply test/quantization/fx/test_quantize_fx.py:4106]: Ok, will do

File: test/quantization/fx/test_quantize_fx.py (hunk 16 above)
Comment (andrewor14) [reply test/quantization/fx/test_quantize_fx.py:4170]: This was actually tested in `test_static_lstm`. This test is only intended for consuming tuples (x[1] is a tuple but x[0] is not)

File: torch/ao/quantization/fx/prepare.py (hunk 8 above)
Comment (andrewor14) [reply torch/ao/quantization/fx/prepare.py:438]: Ok, will do

File: torch/ao/quantization/fx/prepare.py [hunk 18]
```diff
@@ -433,7 +435,13 @@ def get_arg_target_dtype_as_output(
...
+    # to insert DeQuantStubs (see `_insert_dequant_stubs_for_custom_module_lstm_output`)
+    # Here we must trace through that subgraph to reach the original LSTM node
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/prepare.py:440]: I think changing "trace through ..." to "trace back from the args to reach ..." might be more accurate

**Comment by jerryzh168** [comment]
can we write down all the special cases we make for lstm in prepare and convert step? this will help us to evaluate how well this is aligned with the long term plan

File: torch/ao/quantization/fx/utils.py (hunk 17 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/utils.py:822]: oh I see, makes sense

File: torch/ao/quantization/fx/prepare.py [hunk 19]
```diff
@@ -1256,26 +1273,27 @@ def insert_observers_for_model(
...
-                            #   cur_node -> obs
-                            #
-                            # We need to save orig users before updating uses because
-                            # the list of users will change as we update uses
-                            orig_users = list(node.users.keys())
-                            for user_node in orig_users:
-                                if user_node is maybe_output_obs_node:
-                                    continue
-                                user_node.replace_input_with(node, maybe_output_obs_node)
+                            if not _is_custom_module_lstm(node, modules):
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/ao/quantization/fx/prepare.py:1276]: also would be good to put in comments saying here we assume that for quantizable lstm module we assume the output is observed internally. since we'll revisit this when we want to properly support this

File: torch/ao/quantization/fx/utils.py [hunk 20]
```diff
@@ -640,3 +641,220 @@ def get_skipped_module_name_and_classes(
...
+      lstm_output[0] -> output_dq
+      lstm_output[1] -> hidden_dq
+      lstm_output[1][0] -> hidden0_dq
+      lstm_output[1][1] -> hidden1_dq
+
+    Return the node `lstm_output_dq`.
+    """
+    # (1) Split the LSTM node into (output, (hidden0, hidden1))
+    # (2) Insert a DeQuantStub after each internal node
+    original_lstm_users = list(node.users.keys())
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/utils.py:759]: looks like this is not used?

File: torch/ao/quantization/fx/utils.py [hunk 21]
```diff
@@ -640,3 +641,220 @@ def get_skipped_module_name_and_classes(
...
+                    node_to_replacement[hidden_user] = hidden0_dq  # lstm_output[1][0]
+                if _is_getitem_node(hidden_user) and hidden_user.args[1] == 1:
+                    node_to_replacement[hidden_user] = hidden1_dq  # lstm_output[1][1]
+    # Do not reroute nodes we added or the original nodes that will be consumed
+    nodes_added = [output, output_dq, hidden, hidden_dq, hidden0, hidden0_dq, hidden1, hidden1_dq]
+    nodes_to_skip = set(node_to_replacement.keys()).union(nodes_added)
+    for original_node, replacement in node_to_replacement.items():
+        for user in list(original_node.users.keys()):
+            if user not in nodes_to_skip:
+                user.replace_input_with(original_node, replacement)
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/utils.py:797]: do we need to do this? is it enough to just update all uses of `lstm_output_dq`?

Comment (jerryzh168) [reply torch/ao/quantization/fx/utils.py:797]: I guess it might break the reference quantized pattern matching, but feels like we should update the matching code instead

Comment (jerryzh168) [reply torch/ao/quantization/fx/utils.py:797]: or have some optimization passes, e.g.:
```
x --> c = tuple(x, ...) -> c[0] --> output
```
--->

```
x --> output
```

File: torch/ao/quantization/fx/convert.py [hunk 22]
```diff
@@ -491,27 +510,30 @@ def convert_custom_module(
...
+            assert isinstance(hidden0, Node)
+            assert isinstance(hidden1, Node)
+            _remove_previous_dequantize_in_custom_module(node, inputs, graph)
+            _remove_previous_dequantize_in_custom_module(node, hidden0, graph)
+            _remove_previous_dequantize_in_custom_module(node, hidden1, graph)
+        else:
+            # remove the previous dequant node to ensure the inputs are quantized
+            arg = node.args[0]
+            assert isinstance(arg, Node)
+            _remove_previous_dequantize_in_custom_module(node, arg, graph)
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/convert.py:532]: maybe as a followup, we could also remove these code and instead in the prepare code, we add a QuantStub for the inputs that is consumed by lstm nodes, and in convert pass, we can just convert the QuantStub to quantize_per_tensor, this would be more aligned with long term plan I think

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
looks good overall, had some nit comments inline, and also saw some comments from Vasiliy that are not addressed yet, I can do another review after all comments are addressed

File: torch/ao/quantization/fx/convert.py (hunk 22 above)
Comment (vkuzo) [reply torch/ao/quantization/fx/convert.py:532]: It would also be good to value consistency with the current codebase, this would mean making this PR follow what the current codebase is doing (which is inserting observers and converting those to quant/dequant), and not include the "add qstub/dqstub" change into this PR.

Since both this change and the "add qstub/dqstub to FX graph mode quantization" are large, risky, and unrelated, it would be good to keep them separate.  Both of them run some risk of taking more time than anticipated, and if that happens it would be good to have less half-finished branches in the codebase.

Comment (andrewor14) [reply torch/ao/quantization/fx/convert.py:532]: I agree consistency is important. However, using DeQuantStub here actually also significantly simplifies the matching logic in convert, otherwise we would have to do some additional matching to find the "special" observers, which adds some complexity. If you feel strongly about this then I can make the change, otherwise I would prefer to leave this.

File: torch/ao/quantization/fx/utils.py (hunk 21 above)
Comment (andrewor14) [reply torch/ao/quantization/fx/utils.py:797]: Yes, it would break the reference pattern matching. Also, redirecting just `lstm_output_dq` would add some unnecessary nodes for simple use cases like `lstm[0]`. I feel changing the matching logic is a much bigger scope than this task. Maybe we should do it separately?

Comment (jerryzh168) [reply torch/ao/quantization/fx/utils.py:797]: how about this suggestion: https://github.com/pytorch/pytorch/pull/85068#discussion_r975889634

Comment (andrewor14) [reply torch/ao/quantization/fx/utils.py:797]: As discussed offline, we will have an additional pass in this PR to match the pattern `tuple -> tuple -> getitem -> getitem`, for an arbitrary number of `tuple/getitem` pairs

File: torch/ao/quantization/fx/utils.py (hunk 11 above)
Comment (andrewor14) [reply torch/ao/quantization/fx/utils.py:652]: added a check for the qconfig and the qhandler

File: torch/ao/quantization/fx/convert.py (hunk 22 above)
Comment (vkuzo) [reply torch/ao/quantization/fx/convert.py:532]: If we are going to break consistency, we should clearly document why we are doing this, and whether we expect the break in consistency to go away in the future or if it needs to stay around forever.  Can we add this documentation to both `prepare` and `convert` code?

Comment (andrewor14) [reply torch/ao/quantization/fx/convert.py:532]: Sounds good, will do

**Comment by andrewor14** [comment]
> can we write down all the special cases we make for lstm in prepare and convert step? this will help us to evaluate how well this is aligned with the long term plan

Sure. In prepare, we make two special cases:

1. When inserting output observers, we instead break the LSTM output tuple into its internal nodes, insert a DeQuantStub after each node, and recombine the tuple according to the original format ([code](https://github.com/pytorch/pytorch/blob/49cc26784b58fed1a2bd1bfafc31e7a80f668b17/torch/ao/quantization/fx/prepare.py#L1321)). This diverges from, but is conceptually the same as, how we insert "special" output observers that will be converted to "dequantize" nodes (during convert) today for other custom modules. In the future, we should just insert DeQuantStubs instead of these "special" observers for custom modules in general.

2. When inserting input observers for a node, we check whether it is a consumer of LSTM by traversing up the subgraph created in (1) ([code](https://github.com/pytorch/pytorch/blob/49cc26784b58fed1a2bd1bfafc31e7a80f668b17/torch/ao/quantization/fx/prepare.py#L476)). Once we identified what the previous node is, we reuse the [existing code](https://github.com/pytorch/pytorch/blob/49cc26784b58fed1a2bd1bfafc31e7a80f668b17/torch/ao/quantization/fx/prepare.py#L596) to decide whether or not to insert the input observer based on the output dtype of the previous node.

As for convert:

1. When converting the custom module, we change `quantize - dequantize - custom_module` to `quantize - custom_module` for all inputs of custom module LSTM, including the hidden inputs ([code](https://github.com/pytorch/pytorch/blob/49cc26784b58fed1a2bd1bfafc31e7a80f668b17/torch/ao/quantization/fx/convert.py#L525-L527)). This is consistent with how we do the same for custom module inputs today, except it handles more than one input.

2. We replace DeQuantStubs with "dequantize" nodes. This is not really the special case as we plan to do this for all custom modules in the future.

File: torch/ao/quantization/fx/prepare.py [hunk 23]
```diff
@@ -1295,10 +1305,26 @@ def insert_observers_for_model(
...
+                            # LSTM specifically, the outputs are also a nested tuple, so we must first
+                            # break down the tuple to insert DeQuantStubs after the internal nodes.
+
+                            # TODO: This currently diverges from how custom modules are handled today,
+                            # where we insert observers after the output instead of DeQuantStubs, and
+                            # replace these observers with "dequantize" nodes during convert. Conceptually,
+                            # these output observers are the same as DeQuantStubs. In the future, we
+                            # should resolve this inconsistency by inserting DeQuantStubs for all custom
+                            # modules, not just for LSTM.
+                            _insert_dequant_stubs_for_custom_module_lstm_output(node, model, modules, graph)
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/prepare.py:1321]: it feels like `_insert_dequant_stubs_for_custom_module_lstm_output` is not doing equivalent things as `maybe_insert_output_observer_for_node`, e.g. `maybe_output_obs_node` will go through L1328-1348 to update the user node to use this, but looks like this is done inside `_insert_dequant_stubs_for_custom_module_lstm_output`
I'm wondering if branching is placed in a wrong place, should the branch be one level outer?

File: torch/ao/quantization/fx/utils.py [hunk 24]
```diff
@@ -642,3 +646,299 @@ def get_skipped_module_name_and_classes(
...
+        [match_tuple, match_dq, match_getitem, match_lstm],
+    ]
+
+    for p in all_match_patterns:
+        matched_node = _match_pattern(p)
+        if matched_node is not None:
+            return matched_node
+    return None
+
+def _reroute_tuple_getitem_pattern(graph: Graph):
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/utils.py:860]: cc @SherlockNoMad do we want to support something similar in some default fx optimization pass?

File: torch/ao/quantization/fx/utils.py [hunk 25]
```diff
@@ -642,3 +646,299 @@ def get_skipped_module_name_and_classes(
...
+            |
+            d
+
+    After:
+
+        b
+        |
+        d
+    """
+    def find_patterns(
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/utils.py:887]: high level comment is that I think it might be easier if we find pattern in a backward order, since if we go from output to input, there can be only one path, this will be a full graph optimization that can be implemented outside of the `_insert_dequant_stubs_for_custom_module_lstm_output` though. I'll accept to unblock, but if we need to implement this for more general cases I think it would make sense to do it that way.

**Review by jerryzh168 — APPROVED** [review]
looks good, thanks for the hard work!

**Comment by jerryzh168** [comment]
> > can we write down all the special cases we make for lstm in prepare and convert step? this will help us to evaluate how well this is aligned with the long term plan
> 
> Sure. In prepare, we make two special cases:
> 
> 1. When inserting output observers, we instead break the LSTM output tuple into its internal nodes, insert a DeQuantStub after each node, and recombine the tuple according to the original format ([code](https://github.com/pytorch/pytorch/blob/49cc26784b58fed1a2bd1bfafc31e7a80f668b17/torch/ao/quantization/fx/prepare.py#L1321)). This diverges from, but is conceptually the same as, how we insert "special" output observers that will be converted to "dequantize" nodes (during convert) today for other custom modules. In the future, we should just insert DeQuantStubs instead of these "special" observers for custom modules in general.
> 2. When inserting input observers for a node, we check whether it is a consumer of LSTM by traversing up the subgraph created in (1) ([code](https://github.com/pytorch/pytorch/blob/49cc26784b58fed1a2bd1bfafc31e7a80f668b17/torch/ao/quantization/fx/prepare.py#L476)). Once we identified what the previous node is, we reuse the [existing code](https://github.com/pytorch/pytorch/blob/49cc26784b58fed1a2bd1bfafc31e7a80f668b17/torch/ao/quantization/fx/prepare.py#L596) to decide whether or not to insert the input observer based on the output dtype of the previous node.
> 
> As for convert:
> 
> 1. When converting the custom module, we change `quantize - dequantize - custom_module` to `quantize - custom_module` for all inputs of custom module LSTM, including the hidden inputs ([code](https://github.com/pytorch/pytorch/blob/49cc26784b58fed1a2bd1bfafc31e7a80f668b17/torch/ao/quantization/fx/convert.py#L525-L527)). This is consistent with how we do the same for custom module inputs today, except it handles more than one input.
> 2. We replace DeQuantStubs with "dequantize" nodes. This is not really the special case as we plan to do this for all custom modules in the future.

thanks, please add this to the Summary as well

File: torch/ao/quantization/fx/prepare.py (hunk 23 above)
Comment (andrewor14) [reply torch/ao/quantization/fx/prepare.py:1321]: I agree. The only code that we still need for the LSTM case is [these two lines](https://github.com/pytorch/pytorch/blob/4092c24e234ebb2c4ff5eb40b4eb3ebe187003c2/torch/ao/quantization/fx/prepare.py#L1361-L1362):
```
if qhandler is not None and qhandler.is_custom_module():
    swap_custom_module_to_observed(node, qconfig, modules, prepare_custom_config)
```
Maybe I should just copy the second line (the swap) to the LSTM case?

File: torch/ao/quantization/fx/utils.py (hunk 25 above)
Comment (andrewor14) [reply torch/ao/quantization/fx/utils.py:887]: Sounds good, I'll add a TODO

File: torch/ao/quantization/fx/prepare.py (hunk 23 above)
Comment (andrewor14) [reply torch/ao/quantization/fx/prepare.py:1320]: Ok, I copied that line over to the LSTM case and put everything else in the else case

**Comment by andrewor14** [comment]
Ok, I'm merging this. Thanks for all the feedback!
